### PR TITLE
home function

### PR DESCRIPTION
**Pull request by weipinglu**

### CODE REVIEW

File: src/Potato/Data/Text/Zipper2.hs [hunk 1]
```diff
@@ -76,6 +76,12 @@ mapZipper f (TextZipper lb b s a la) = TextZipper
   , _textZipper_linesAfter = fmap (T.map f) la
   }
 
+appendEnd :: [Text] -> Text -> [Text]
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs:81]: very easy to read this function and think it's [snoc](https://hackage.haskell.org/package/extra-1.7.12/docs/Data-List-Extra.html#v:snoc)

maybe give it a more descriptive name

File: src/Potato/Data/Text/Zipper2.hs [hunk 2]
```diff
@@ -154,7 +156,11 @@ pageDown pageSize z = undefined
 
 -- | Move the cursor to the beginning of the current logical line (clearing the selection)
 home :: TextZipper -> TextZipper
-home (TextZipper lb b s a la) = undefined
+home (TextZipper lb b [] a la) = TextZipper lb "" [] (b <> a) la
+home (TextZipper lb b (x:[]) a la) = TextZipper lb "" [] (b <> x) (a:la)
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs]: I don't think this case is correct. What if your zipper looks like this ([] is where the cursor (emptyselection) is)

`what happens if [] you go to home of this line`

**Comment by minimapletinytools** [comment]
how about adding a UT for `home`?

File: src/Potato/Data/Text/Zipper2.hs (hunk 2 above)
Comment (minimapletinytools) [reply src/Potato/Data/Text/Zipper2.hs]: original:
this is an example
this line is selected [] lines after
blah blah blah


correct:
this is an example
[]this line is selected  lines after
blah blah blah


I think this is what you have:
this is an example
[]this line is selected 
 lines after
blah blah blah

Comment (weipinglu) [reply src/Potato/Data/Text/Zipper2.hs]: It does take care of this case. Please see the test: "home example"

Comment (minimapletinytools) [reply src/Potato/Data/Text/Zipper2.hs]: ah my example is not right then. It should look like this



original:
```
this is an example
this line is selected [pooh bear] lines after
blah blah blah
```

correct:
```
this is an example
[]this line is selected pooh bear lines after
blah blah blah
```

I think this is what you have:
```
this is an example
[]this line is selected
pooh bear lines after
blah blah blah

```

Comment (weipinglu) [reply src/Potato/Data/Text/Zipper2.hs]: fixed

File: src/Potato/Data/Text/Zipper2.hs [hunk 3]
```diff
@@ -154,7 +156,10 @@ pageDown pageSize z = undefined
 
 -- | Move the cursor to the beginning of the current logical line (clearing the selection)
 home :: TextZipper -> TextZipper
-home (TextZipper lb b s a la) = undefined
+home (TextZipper lb b [] a la) = TextZipper lb "" [] (b <> a) la
+home (TextZipper lb b (x:[]) a la) = TextZipper lb "" [] (b <> x <> a) la
+--home (TextZipper lb b (x:xs) a la) = case xs of
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs]: you can delete commented line here

File: src/Potato/Data/Text/Zipper2.hs [hunk 4]
```diff
@@ -154,7 +156,10 @@ pageDown pageSize z = undefined
...
+home (TextZipper lb b (x:xs) a la) = TextZipper lb "" [] (b <> x) ((init xs) <> [(last xs) <> a] <> la)
```
Comment (minimapletinytools) [inline src/Potato/Data/Text/Zipper2.hs]: The pattern matching guarantees xs is non-empty so `last` is fine here, but it's still a [partial function](https://wiki.haskell.org/Partial_functions)

The way to encode the pattern matching non-emptyness guarantee of xs being non empty would be to use [NonEmpty](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-List-NonEmpty.html) which is constructed using the `(:|) :: a -> [a] -> NonEmpty a` operator. Then you could do something like:
```
home (TextZipper lb b (x:(xs:xss)) a la) = TextZipper lb "" [] (b <> x) ((init xs') <> [(last xs') <> a] <> la) where
   xs' = xs :| xss
```

That's why I went through the trouble of writing appendEnd,  but using NonEmpty might actually be better!

Comment (minimapletinytools) [reply src/Potato/Data/Text/Zipper2.hs]: Note both Prelude and NonEmpty exports init/last so you need to qualify them i.e.

import Data.List.NonEmpty as NE
NE.init/NE.last

**Review by minimapletinytools — APPROVED** [review]

**Review by minimapletinytools — APPROVED** [review]

**Pull request merged**